sqlbase: multirowfetcher stops decoding a row when it doesn't need more data

jordanlewis opened the pull request:
Previously, if the row fetcher needed any of the columns from the value
portion of the keys in a row, it would iterate over the whole value
column by column, decoding the needed columns and skipping unneeded
columns by reading their length. This would occur even after all of the
needed columns were decoded.

This is inefficient for certain kinds of tables and queries. For
example, consider a very wide table and a query that only needs values
from the beginning of the table. In that case, the row fetcher would
read the length of every column value even once it had gotten all of the
column values that it needed.

After this change, the row fetcher keeps track of the number of column
values it needs, and quits decoding early once it has read all of the
required columns. The same is done in `finalizeRow`, which now quits
early when all needed columns are set.

```
name                                old time/op    new time/op    delta
WideTableIgnoreColumns/Cockroach-8    21.5ms ± 2%    14.8ms ± 2%  -31.09%  (p=0.000 n=10+10)

name                                old alloc/op   new alloc/op   delta
WideTableIgnoreColumns/Cockroach-8    2.14MB ± 1%    2.12MB ± 0%   -1.02%  (p=0.000 n=10+9)

name                                old allocs/op  new allocs/op  delta
WideTableIgnoreColumns/Cockroach-8     1.41k ±54%     1.33k ±29%     ~     (p=0.239 n=10+10)
```

Release note (performance improvement): Queries that only need to read
part of a table with many columns are more efficient.

Fixes #21443.

jordanlewis committed:
sqlbase: remove unused argument to DecodeKeyVals

Release note: None

jordanlewis committed:
sqlbase: stop decoding vals when possible

Previously, if the row fetcher needed any of the columns from the value
portion of the keys in a row, it would iterate over the whole value
column by column, decoding the needed columns and skipping unneeded
columns by reading their length. This would occur even after all of the
needed columns were decoded.

This is inefficient for certain kinds of tables and queries. For
example, consider a very wide table and a query that only needs values
from the beginning of the table. In that case, the row fetcher would
read the length of every column value even once it had gotten all of the
column values that it needed.

After this change, the row fetcher keeps track of the number of column
values it needs, and quits decoding early once it has read all of the
required columns.

```
name                                old time/op    new time/op    delta
WideTableIgnoreColumns/Cockroach-8    21.7ms ± 2%    15.7ms ± 2%  -27.58%  (p=0.000 n=10+9)

name                                old alloc/op   new alloc/op   delta
WideTableIgnoreColumns/Cockroach-8    2.14MB ± 1%    2.12MB ± 0%   -1.09%  (p=0.000 n=10+10)

name                                old allocs/op  new allocs/op  delta
WideTableIgnoreColumns/Cockroach-8     1.43k ±63%     1.35k ±28%     ~     (p=0.436 n=10+10)
```

Release note (performance improvement): Queries that only need to read
part of a table with many columns are more efficient.

jordanlewis committed:
bench: add benchmark for select from wide table

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21450)
<!-- Reviewable:end -->

jordanlewis committed:
sqlbase: rowfetcher's finalizeRow does less work

Previously, MultiRowFetcher.finalizeRow looped over every table column,
checked if it was needed but not found in the row, and set such a column
to NULL for presentation.

This is not wrong, but it's inefficient - we don't need to do any work
if we already know we've gotten all of the columns that we need. Tables
with many columns cause this loop to take appreciable time, because it's
not free to check whether a column is in the needed set.

Now, we quit the loop early when we detect that we've gotten all needed
columns.

```
name                                old time/op    new time/op    delta
WideTableIgnoreColumns/Cockroach-8    16.7ms ± 2%    15.5ms ± 2%  -7.23%  (p=0.000 n=10+10)

name                                old alloc/op   new alloc/op   delta
WideTableIgnoreColumns/Cockroach-8    2.13MB ± 0%    2.12MB ± 0%  -0.26%  (p=0.004 n=10+10)

name                                old allocs/op  new allocs/op  delta
WideTableIgnoreColumns/Cockroach-8     1.51k ±22%     1.39k ±25%    ~     (p=0.086 n=10+10)
```

Release note (performance improvement): Queries on tables with many
columns are more efficient.

jordanlewis commented:
Added another commit that does a similar optimization in `finalizeRow`, which previously looped over all rows as well. It gave another 5% improvement. I updated the performance numbers in the PR message to include both commits.

RaduBerinde reviewed (approved):
LGTM

In the last commit message, I think you mean columns ("if we already know we've gotten all of the rows that we need").

jordanlewis commented:
TFTR, yeah I said rows instead of columns twice here. Thanks!

jordanlewis commented on pkg/sql/sqlbase/multirowfetcher.go:
```diff
@@ -1093,6 +1097,7 @@ func (mrf *MultiRowFetcher) finalizeRow() error {
 			table.row[i] = EncDatum{
 				Datum: tree.DNull,
 			}
+			mrf.valueColsFound++
```
Done.
